### ISSUE TITLE
Accuracy Drop in Distributed Training of Astroformer on CIFAR-10

### ISSUE THREAD

**Issue by 181404010226** [Rishit-dagli/Astroformer#5]
I successfully reproduced Astroformer's performance on CIFAR-10, achieving over 98% accuracy. However, I noticed a significant drop in model accuracy when using distributed training. This could potentially explain why Astroformer-5 only achieved 93.23% accuracy. Is there a known issue with distributed training for this architecture?

**Comment by sjtuccz** [comment]
Hello, did you use torchvision to load the cifa dataset and what size was used as input? 224 or the original 32?Do you use mixup?

**Comment by 181404010226** [comment]
Perhaps this is no longer a significant issue, as single-GPU training can achieve the accuracy reported in the paper, with the use of mixup and the image size already expanded to 224.



---Original---
From: ***@***.***&gt;
Date: Tue, Feb 4, 2025 19:51 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [Rishit-dagli/Astroformer] Accuracy Drop in Distributed Trainingof Astroformer on CIFAR-10 (Issue #5)




 
Hello, did you use torchvision to load the cifa dataset and what size was used as input? 224 or the original 32?Do you use mixup?

**Comment by sjtuccz** [comment]
Thanks for your answer. But training astroformer3 from scratch with cifar100 did not achieve the 87.65 accuracy rate mentioned in the article . I got a result by a wide margin( 78.46). Moreover, adjusting the learning rate has little effect on the accuracy.

CUDA_VISIBLE_DEVICES=3 python train.py --dataset torch/cifar100 --data-dir pytorch-cifar100-master/data --model astroformer_3 --num-classes 100 --in-chans 3 --input-size 3 224 224 --batch-size 128 --grad-accum-steps 1 --opt adamw --sched cosine --lr-base 2e-5 --lr-cycle-decay 1e-2 --lr-k-decay 1 --warmup-lr 1e-5 --epochs 300 --warmup-epochs 5 --mixup 0.8 --smoothing 0.1 --drop 0.1 --amp --output /Astroformer-main/output

![Image](https://github.com/user-attachments/assets/a07f352b-4aea-4acb-a4c3-f9c8d1de9f56)

**Comment by sjtuccz** [comment]
How is it different from the parameters you used? How did you achieve such high accuracy? I am looking forward to your reply,thanks a lot.

**Comment by sjtuccz** [comment]
The accuracy on cifar10 is about 95%-96%, and it is basically impossible to increase to 99.12% after training
![Image](https://github.com/user-attachments/assets/2196d126-5d87-4e11-9eee-2322aefea847)

**Comment by 181404010226** [comment]
You are correct; training directly on CIFAR-100 might not yield optimal results. The mentioned performance could likely be achieved through transfer learning. However, the results on CIFAR-10 are verifiable (Astroformer0 can achieve an accuracy of 97.89%), which is why I did not delve deeper into proving it further. If you are interested, I suggest you could experiment with ImageNet first and then transfer the learning to CIFAR-100.



---Original---
From: ***@***.***&gt;
Date: Thu, Feb 6, 2025 11:21 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [Rishit-dagli/Astroformer] Accuracy Drop in Distributed Trainingof Astroformer on CIFAR-10 (Issue #5)




 
Thanks for your answer. But training astroformer3 from scratch with cifar100 did not achieve the 87.65 accuracy rate mentioned in the article . I got a result by a wide margin( 78.46). Moreover, adjusting the learning rate has little effect on the accuracy.
 
CUDA_VISIBLE_DEVICES=3 python train.py --dataset torch/cifar100 --data-dir pytorch-cifar100-master/data --model astroformer_3 --num-classes 100 --in-chans 3 --input-size 3 224 224 --batch-size 128 --grad-accum-steps 1 --opt adamw --sched cosine --lr-base 2e-5 --lr-cycle-decay 1e-2 --lr-k-decay 1 --warmup-lr 1e-5 --epochs 300 --warmup-epochs 5 --mixup 0.8 --smoothing 0.1 --drop 0.1 --amp --output /Astroformer-main/output
 
default.png (view on web)

**Comment by 181404010226** [comment]
@sjtuccz I did not use the built-in parameters for training; instead, I utilized my own data augmentation library. You can refer to the parameters in my code to conduct the training.

`import os
import torch
import numpy as np
import random
from timm.data import create_loader, Mixup
from torchvision import datasets, transforms
from Paper_global_vars import global_vars
from torchvision.transforms import Resize
from pprint import pprint
import torch.distributed as dist
from torch.utils.data.distributed import DistributedSampler
# 使用 functools.partial 创建一个可序列化的 collate 函数
from functools import partial
   
# 设置随机种子
seed = 42
torch.manual_seed(seed)
np.random.seed(seed)
random.seed(seed)
if torch.cuda.is_available():
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)  # if you are using multi-GPU.
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False


IMAGENET_DEFAULT_MEAN = (0.485, 0.456, 0.406)
IMAGENET_DEFAULT_STD = (0.229, 0.224, 0.225)

cifar10_mean = (0.4914, 0.4822, 0.4465)
cifar10_std = (0.2471, 0.2435, 0.2616)

# ImageNet数据集路径
IMAGENET_ROOT = '/root/autodl-tmp/imagenet'

# 定义数据配置
data_config = {
    'input_size':global_vars.input_size,
    'interpolation': 'bicubic',
    'mean': IMAGENET_DEFAULT_MEAN,
    'std': IMAGENET_DEFAULT_STD,
    'crop_pct': 0.96,
}

mixup_args = dict(
    mixup_alpha=0.5,
    cutmix_alpha=0.5,
    prob=1.0,
    switch_prob=0.5,
    mode='batch',
    label_smoothing=0.1,
    num_classes=100
)

# 将 collate_mixup_fn 移到顶级作用域
def collate_mixup_fn(batch, mixup_fn):
    inputs = torch.stack([b[0] for b in batch])
    targets = torch.tensor([b[1] for b in batch])
    return mixup_fn(inputs, targets)

# 选择数据集
root = os.path.join(os.path.dirname(__file__), "CIFAR10RawData")
# 修改创建训练数据加载器的部分
def create_train_loader(dataset='cifar10', distributed=False):
    global loader_train
    
    # 为不同数据集定义最优的增强配置
    augmentation_config = {
        'cifar10': {
            'scale': (0.75, 1.0),
            're_prob': 0.25,
            'auto_augment': 'rand-m9-mstd0.5-inc1',
            'color_jitter': 0.4,
            'mean': cifar10_mean,
            'std': cifar10_std,
        },
        'cifar100': {
            'scale': (0.75, 1.0),
            're_prob': 0.25,
            'auto_augment': 'rand-m9-mstd0.5-inc1',
            'color_jitter': 0.4,
            'mean': cifar10_mean,  
            'std': cifar10_std,
        },
        'imagenet': {
            'scale': (0.08, 1.0),
            're_prob': 0.5,
            'auto_augment': 'rand-m9-mstd0.5-inc1',
            'color_jitter': 0.4,
            'mean': IMAGENET_DEFAULT_MEAN,
            'std': IMAGENET_DEFAULT_STD,
        }
    }

    # 选择数据集
    if dataset == 'cifar10':
        trainset = datasets.CIFAR10(root=root, train=True, download=True, transform=None)
        num_classes = 10
    elif dataset == 'cifar100':  
        trainset = datasets.CIFAR100(root=root, train=True, download=True, transform=None)
        num_classes = 100
    elif dataset == 'imagenet':
        trainset = datasets.ImageNet(root=IMAGENET_ROOT, split='train')
        num_classes = 1000
    else:
        raise ValueError("Invalid dataset. Choose 'cifar10', 'cifar100', 'imagenet'")
    
    # 更新mixup配置
    mixup_args['num_classes'] = num_classes
    mixup_fn = Mixup(**mixup_args)
    collate_fn = partial(collate_mixup_fn, mixup_fn=mixup_fn)
    
    # 获取当前数据集的增强配置
    aug_config = augmentation_config[dataset]
    
    loader_train = create_loader(
        trainset,
        input_size=data_config['input_size'],
        batch_size=global_vars.train_batch_size,
        is_training=True,
        use_prefetcher=False,
        no_aug=False,
        re_prob=aug_config['re_prob'],
        re_mode='pixel',
        re_count=1,
        scale=aug_config['scale'],
        ratio=(3./4., 4./3.),
        hflip=0.5,
        vflip=0.,
        color_jitter=aug_config['color_jitter'],
        auto_augment=aug_config['auto_augment'],
        num_aug_splits=0,
        interpolation=data_config['interpolation'],
        mean=aug_config['mean'],
        std=aug_config['std'],
        num_workers=8,
        distributed=distributed,
        crop_pct=data_config['crop_pct'],       
        collate_fn=collate_fn,
        use_multi_epochs_loader=False,
        worker_seeding='all',  
        pin_memory=True
    )
    return loader_train

# 修改创建验证数据加载器的部分
def create_valid_loader(dataset='cifar10', distributed=False):
    global valid_data
    
    if dataset == 'cifar10':
        testset_cifar10 = datasets.CIFAR10(root=root, train=False, download=True, transform=None)
        testset = testset_cifar10
    elif dataset == 'cifar100':
        testset_cifar100 = datasets.CIFAR100(root=root, train=False, download=True, transform=None)
        testset = testset_cifar100
    elif dataset == 'imagenet':
        testset = datasets.ImageNet(root=IMAGENET_ROOT, split='val')
    else:
        raise ValueError("Invalid dataset. Choose 'cifar10', 'cifar100', 'imagenet'")
    
    valid_data = create_loader(
        testset,
        input_size=data_config['input_size'],
        batch_size=global_vars.test_batch_size,
        is_training=False,
        use_prefetcher=False,
        interpolation=data_config['interpolation'],
        mean=data_config['mean'],
        std=data_config['std'],
        num_workers=8,
        distributed=distributed,
        crop_pct=data_config['crop_pct'],
        pin_memory=True
    )
    return valid_data
`

**Comment by sjtuccz** [comment]
WOW let me try.非常感谢
